### PR TITLE
Fix #659 AxiMemorySim for AXI4 specification

### LINKED ISSUE #659: Fix `AxiMemorySim`?
I found that `spinal.lib.bus.amba4.axi.sim.AxiMemorySim` does not obey AXI4 specification
There are a few sub-issues
1. Cannot correctly reply `rid` (not driven now)
2. read/write transactions is not correct when address is not aligned
e.g. 
We have `32'h3210` at address 0 and `32'h7654` at address 4.
The bus should reply `32'h3210` (generally `32'h--1-`) when address 1 is required and `arsize` is set to 0, instead of `32'h4321`

### PR DESCRIPTION
**Pull request by tanhongze**
1. support narrow transfer

2. support burst and write strobes

3. support id replying (arid->rid,awid->bid)

4. support read and write response

5. check burst does not cross 4KB boundary

### CODE REVIEW

**Comment by Readon** [comment]
Good work.
Is there any possibility that using Axi4Agent components to construct the Axi4MemorySim so that the code for spec compatibility can be handled only in one group of classes but reimplement them in several files?

**Comment by tanhongze** [comment]
> Good work. Is there any possibility that using Axi4Agent components to construct the Axi4MemorySim so that the code for spec compatibility can be handled only in one group of classes but reimplement them in several files?

I don't know. I probably do not have enough time to think how to improve this in a few weeks.

**Comment by Dolu1990** [comment]
@c-thaler This update AxiMemorySim, good to merge ?

**Comment by tanhongze** [comment]
Oh, sorry. I lost my last modification in this PR.

**Comment by tanhongze** [comment]
It should be OK now. 
Should I provide my tester somewhere?

**Comment by c-thaler** [comment]
Besides the wrong error message I commented on in the code, I'm fine with it.

**Review by c-thaler — CHANGES_REQUESTED** [review]

File: lib/src/main/scala/spinal/lib/bus/amba4/axi/sim/AxiMemorySim.scala [hunk 1]
```diff
@@ -210,9 +226,40 @@ case class SparseMemory() {
...
+    if(ret >= upperWrapBoundary){
+      ret - dataTransactionSize
+    }
+    else ret
+  }
   // check for read/write over 4k boundary
+  if(burstType == 1){
+    assert(
+      assertion = ((burstLength << burstSize) + (address & 4095)) <= 4095,
+      message   = s"Write request crossing 4k boundary (addr=${address.toHexString}, len=${burstLength.toLong.toHexString}"
```
[the remark this answers is not in the document]
Comment (tanhongze) [reply lib/src/main/scala/spinal/lib/bus/amba4/axi/sim/AxiMemorySim.scala]: You are right. I will fix it.

**Comment by Dolu1990** [comment]
> Should I provide my tester somewhere?

The tester is welcome, as long as it take less than 10 seconds to run ^^

**Comment by Dolu1990** [comment]
@tanhongze Do you still have the tests ?

**Comment by tanhongze** [comment]
> @tanhongze Do you still have the tests ?

My tests are too long to run now. Smarter tests without recompiling are needed, I guess. 
Maybe I will put it in the next PR after I rewrite it someday. Is that OK?

**Comment by Dolu1990** [comment]
All good, thanks ^^

**Pull request merged**